### PR TITLE
[lldb/crashlog] Fix module binary resolution

### PR DESCRIPTION
**Pull request by medismailben**
This patch fixes a bug in when resolving and loading modules from the binary image list.

When loading a module, we would first use the UUID from the binary image list with `dsymForUUID` to fetch the dSYM bundle from our remote build records and copy the executable locally.

If we failed to find a matching dSYM bundle for that UUID on the build record, let's say if that module was built locally, we use Spotlight (`mdfind`) to find the dSYM bundle once again using the UUID.

Prior to this patch, we would set the image path to be the same as the symbol file. This resulted in trying to load the dSYM as a module in lldb, which isn't allowed.

This patch address that by looking for a binary matching the image identifier, next to the dSYM bundle and try to load that instead.

rdar://127433616

### CODE REVIEW

File: lldb/examples/python/crashlog.py [hunk 1]
```diff
@@ -418,9 +418,22 @@ def locate_module_and_debug_symbols(self):
...
+                                % parent_dir,
+                                shell=True,
+                            )
+                            .decode("utf-8")
+                            .splitlines()
+                        )
+                        for binary in find_results:
+                            abs_path = os.path.abspath(binary)
+                            basename = os.path.basename(binary)
+                            if os.path.exists(abs_path) and basename == self.identifier:
```
Comment (bulbazord) [inline lldb/examples/python/crashlog.py:435]: Do we expect `abs_path` to not exist after the call to `find`?

File: lldb/examples/python/crashlog.py [hunk 2]
```diff
@@ -418,9 +418,22 @@ def locate_module_and_debug_symbols(self):
...
                             print('falling back to binary inside "%s"' % dsym)
                         self.symfile = dsym
-                        for filename in os.listdir(dwarf_dir):
-                            self.path = os.path.join(dwarf_dir, filename)
-                            if self.find_matching_slice():
+                        # Look for the executable next to the dSYM bundle.
+                        parent_dir = os.path.dirname(dsym)
+                        find_results = (
+                            subprocess.check_output(
+                                '/usr/bin/find "%s" -type f \( -perm -u=x -o -perm -g=x -o -perm -o=x \)'
```
Comment (JDevlieghere) [inline lldb/examples/python/crashlog.py:425]: Rather than shelling out to find, can you use `os.walk`?

File: lldb/examples/python/crashlog.py [hunk 3]
```diff
@@ -418,9 +418,20 @@ def locate_module_and_debug_symbols(self):
...
-                        for filename in os.listdir(dwarf_dir):
-                            self.path = os.path.join(dwarf_dir, filename)
-                            if self.find_matching_slice():
+                        # Look for the executable next to the dSYM bundle.
+                        parent_dir = os.path.dirname(dsym)
+                        executables = []
+                        for root, dirs, files in os.walk(parent_dir):
+                            for file in files:
+                                abs_path = os.path.join(root, file)
+                                if os.path.isfile(abs_path) and os.access(
```
Comment (bulbazord) [inline lldb/examples/python/crashlog.py:427]: This must necessarily be a file right? It's in the `files` output of `os.walk`. Could it be something else?

File: lldb/examples/python/crashlog.py [hunk 4]
```diff
@@ -418,9 +418,20 @@ def locate_module_and_debug_symbols(self):
                         with print_lock:
                             print('falling back to binary inside "%s"' % dsym)
                         self.symfile = dsym
-                        for filename in os.listdir(dwarf_dir):
-                            self.path = os.path.join(dwarf_dir, filename)
-                            if self.find_matching_slice():
+                        # Look for the executable next to the dSYM bundle.
+                        parent_dir = os.path.dirname(dsym)
+                        executables = []
+                        for root, dirs, files in os.walk(parent_dir):
```
Comment (JDevlieghere) [inline lldb/examples/python/crashlog.py:424]: Let's use `_` instead of `dirs` as it's unused.

**Review by JDevlieghere — APPROVED** [review]

File: lldb/examples/python/crashlog.py (hunk 3 above)
Comment (medismailben) [reply lldb/examples/python/crashlog.py:427]: It could be block device or another file type fwiw.

**Review by bulbazord — APPROVED** [review]

**Pull request merged**